Clarifying Freer documentation about linear continuation (Freer can't discard a continuation)

Playing with Preface in my project Orion, I get some trouble dealing with Freer Monad. For example, let's imagine this simple program using Free:  

### A simple program using Free and Freer

```ocaml 

type 'a effect =
  | Get_home of (string -> 'a)
  | Print of (string * (unit -> 'a))

module Free = struct
  include Make.Free_monad.Over_functor (Make.Functor.Via_map (struct
    type 'a t = 'a effect

    let map f = function
      | Get_home resume -> Get_home (fun s -> f (resume s))
      | Print (message, resume) -> Print (message, fun () -> f (resume ()))
    ;;
  end))

  let get_home = perform (Get_home Fun.id)
  let print message = perform (Print (message, Fun.id))

  let program path =
    let open Monad in
    match path with
    | Some path ->
      let* () = print ("Path is " ^ path) in
      return (Try.ok ())
    | None ->
      let* home = get_home in
      let* () = print ("Path is " ^ home) in
      return (Try.ok ())
  ;;
end
```

The program is pretty simple and it is easy to interpret it : 

```ocaml 
let free () =
  let () = print_endline "free" in
  match
    Home.Free.(
      run
        (function
          | Home.Print (message, resume) ->
            let () = print_endline message in
            resume ()
          | Home.Get_home resume ->
            let home = Sys.getenv "HOME" in
            resume home)
        (program path))
  with
  | Ok () -> exit 0
  | Error _ -> exit 1
```

Let's have a look to his Freer implementation : 

```ocaml

module Freer = struct
  include Make.Freer_monad.Over (struct
    type 'a t = 'a effect
  end)

  let get_home = perform (Get_home Fun.id)
  let print message = perform (Print (message, Fun.id))

  let program path =
    let open Monad in
    match path with
    | Some path ->
      let* () = print ("Path is " ^ path) in
      return (Try.ok ())
    | None ->
      let* home = get_home in
      let* () = print ("Path is " ^ home) in
      return (Try.ok ())
  ;;
end
```

As for Free, the program is pretty simple to be interpreted (with a little bit more noise): 

```ocaml

let freer () =
  let () = print_endline "freer" in
  let runtime =
    let open Home.Freer in
    let interpreter = function
      | Home.Print (message, resume) ->
        let () = print_endline message in
        resume ()
      | Home.Get_home resume ->
        let home = Sys.getenv "HOME" in
        resume home
    in
    { interpreter }
  in
  match Home.Freer.(run runtime (program path)) with
  | Ok () -> exit 0
  | Error _ -> exit 1
;;
```

In this case, Free and Freer seems iso but : 
- it is a little bit more difficult to describe a set of effects as a Free monad (we have to provide a Functor for the effect set) but after the interpretation is pretty simple, just give a function.
- It is a little bit more difficult to describe the interpreter (having a record with existential) but the effect description is more easy, just give the type. 

### Changing the interpretation of Get_home

So for many reason, i'm tempted to go for a Freer. But there is a little subtility. Let's change a little bit the interpretation of Get_home : 

```ocaml 
exception No_home

let free () =
  let () = print_endline "free" in
  match
    Home.Free.(
      run
        (function
          | Home.Print (message, resume) ->
            let () = print_endline message in
            resume ()
          | Home.Get_home resume ->
            (match Sys.getenv_opt "HOME" with
            | Some home -> resume home
            | None -> Preface.Try.error No_home))
        (program path))
  with
  | Ok () -> exit 0
  | Error _ -> exit 1
;;
```

In this interpreter, we decide to change the behaviour of the program, if the value for HOME is empty, the program should discard his interpretation. In free there is not problem with that. But in freer : 

```ocaml
let freer () =
  let () = print_endline "freer" in
  let runtime =
    let open Home.Freer in
    let interpreter = function
      | Home.Print (message, resume) ->
        let () = print_endline message in
        resume ()
      | Home.Get_home resume ->
        (match Sys.getenv_opt "HOME" with
        | Some home -> resume home
        | None -> Preface.Try.error No_home)
    in
    { interpreter }
  in
  match Home.Freer.(run runtime (program path)) with
  | Ok () -> exit 0
  | Error _ -> exit 1
;; 
```
We face to an error : 

```
File "Preface_ex.ml", line 37, characters 6-17:
37 |     { interpreter }
           ^^^^^^^^^^^
Error: This field value has type 'b Preface.Try.t f -> 'b Preface.Try.t
       which is less general than 'a. 'a f -> 'a
make: *** [Makefile:2: build] Error 1
```

Free can discard or not the continuation, Freer not (for quantification reason). 

### Using First Class Modules

Without control flow management, using a free monad is not very useful. See, for example the implementation of the program using High Order Modules : 

```ocaml 
module type RUNTIME = sig
  val get_home : unit -> string
  val print : string -> unit
end

let program (module Runtime : RUNTIME) path =
  match path with
  | Some path ->
    let () = Runtime.print ("Path is " ^ path) in
    Try.ok ()
  | None ->
    let home = Runtime.get_home () in
    let () = Runtime.print ("Path is " ^ home) in
    Try.ok ()
;;
```

and: 

```ocaml 

let homodule () =
  let () = print_endline "High order module" in
  let module Runtime = struct
    let get_home () = Sys.getenv "HOME"
    let print = print_endline
  end
  in
  match Home.program (module Runtime) path with
  | Ok () -> exit 0
  | Error _ -> exit 1
;;
```
Since it is not possible to deal with a particular implementation of get_home, exactly like Freer, First class module seems a slightly better alternative than Freer monad.

###  So?

- Should we drop Freer
- Should we found a way to fix Freer
- Do you have any solution?

--- a/lib/preface_make/free_monad.mli
+++ b/lib/preface_make/free_monad.mli
@@ -52,10 +52,10 @@
       module Store_free = Preface_make.Free_monad.Over_functor (Store.Functor)
     ]}
 
-    {3 Defining an interpeter}
+    {3 Defining a Handler}
 
-    Then we can propose one interpretation using an OCaml side effect for
-    instance, here the side effect is a mutable reference
+    Then we can propose a handler using an OCaml side effect for instance, here
+    the side effect is a mutable reference
 
     {[
       let runStore l = function

--- a/lib/preface_make/freer_monad.ml
+++ b/lib/preface_make/freer_monad.ml
@@ -9,13 +9,14 @@ module Over (Type : Preface_specs.Types.T1) = struct
 
   let perform f = Bind (f, (fun a -> Return a))
 
-  type interpreter = { interpreter : 'a. 'a f -> 'a }
+  type 'a handler = { handler : 'b. ('b -> 'a) -> 'b f -> 'a }
 
   let run f =
     let rec loop_run = function
       | Return a -> a
-      | Bind (intermediate, continuation) ->
-        loop_run (continuation (f.interpreter intermediate))
+      | Bind (intermediate, continue) ->
+        let k x = loop_run (continue x) in
+        f.handler k intermediate
     in
     loop_run
   ;;

--- a/lib/preface_make/freer_monad.mli
+++ b/lib/preface_make/freer_monad.mli
@@ -15,12 +15,7 @@
     In this example we propose a Store where strings can be stored and retrieved
     with a key which is also a string.
 
-    {3 Defining a [Functor]}
-
-    The first piece of this jigsaw should be an Algebraic Data Type (ADT) and
-    it's dedicated [map] function. In a language like Haskell such mapping
-    function can be automatically derived. In [Preface] this is not the case
-    yet.
+    {3 Defining effects}
 
     {[
       (* file: store.ml *)
@@ -38,18 +33,18 @@
 
     {[ module Store_free = Preface_make.Freer_monad.Over (Store) ]}
 
-    {3 Defining an interpeter}
+    {3 Defining a Handler}
 
-    Then we can propose one interpretation using an OCaml side effect for
-    instance, here the side effect is a mutable reference
+    Then we can propose a handler using an OCaml side effect for instance, here
+    the side effect is a mutable reference
 
     {[
-      let runStore l = function
+      let runStore l resume = function
         | Store.Get (k, f) ->
-          f (Option.map snd (List.find_opt (fun (k', _) -> k' = k) !l))
+          resume (f (Option.map snd (List.find_opt (fun (k', _) -> k' = k) !l)))
         | Store.Set (k, v, f) ->
           let () = l := (k, v) :: !l in
-          f ()
+          resume (f ())
       ;;
     ]}
 
@@ -64,19 +59,19 @@
       let set k v = Store_free.perform (Store.Set (k, v, id)))
     ]}
 
-    {3 Building the interpreter}
+    {3 Building the handler}
 
-    Therefor the [Freer_monad] requires an interpeter with the following type:
+    Therefor the [Freer_monad] requires a handler with the following type:
 
-    {[ type interpret = { interpreter : 'a. 'a Store.t -> 'a } ]}
+    {[ type 'a handler = { handler : 'b. ('b -> 'a) -> 'b f -> 'a } ]}
 
-    Such interpreter shoud be created thanks to a function with an embedded let
+    Such handler shoud be created thanks to a function with an embedded let
     which provides the more general type i.e. `'a. 'a Store.t -> 'a`.
 
     {[
       let interpreter l =
         let i = runStore l in
-        Store_free.{ interpreter = i }
+        Store_free.{ handler = i }
       ;;
     ]}
 

--- a/lib/preface_specs/free_monad.mli
+++ b/lib/preface_specs/free_monad.mli
@@ -1,6 +1,6 @@
 (** A [Free_monad] allows you to build a monad from a given functor. Such monad
     is equiped with two additional functions: one dedicated to the creation of a
-    new data and another one for the interpretation. *)
+    new data and another one for the effect handling. *)
 
 (** {1 Structure anatomy} *)
 
@@ -18,7 +18,7 @@ module type CORE = sig
   (** Create a new ['a t] from a ['a f]. *)
 
   val run : ('a f -> 'a) -> 'a t -> 'a
-  (** Execute a given interpret for given data *)
+  (** Execute a given handler for given data *)
 end
 
 (** {1 API} *)

--- a/lib/preface_specs/freer_monad.mli
+++ b/lib/preface_specs/freer_monad.mli
@@ -14,14 +14,14 @@ module type CORE = sig
     | Return : 'a -> 'a t
     | Bind : 'b f * ('b -> 'a t) -> 'a t
 
-  type interpreter = { interpreter : 'a. 'a f -> 'a }
-  (** The interpreter type *)
+  type 'a handler = { handler : 'b. ('b -> 'a) -> 'b f -> 'a }
+  (** The handler type *)
 
   val perform : 'a f -> 'a t
   (** Create a new ['a t] from a ['a f]. *)
 
-  val run : interpreter -> 'a t -> 'a
-  (** Execute a given interpret for given data *)
+  val run : 'a handler -> 'a t -> 'a
+  (** Execute a given handler for given data *)
 end
 
 (** {1 API} *)

--- a/test/preface_examples_test/freer_monad_consoleIO.ml
+++ b/test/preface_examples_test/freer_monad_consoleIO.ml
@@ -12,19 +12,19 @@ let tell x = IO.perform (ConsoleIO.Tell (x, id))
 
 let ask s = IO.perform (ConsoleIO.Ask (s, id))
 
-let runConsoleIO output = function
+let runConsoleIO output sk = function
   | ConsoleIO.Tell (s, k) ->
     let () = output := !output @ [ "Tell " ^ s ] in
-    k ()
+    sk (k ())
   | ConsoleIO.Ask (s, k) ->
     let () = output := !output @ [ "Ask " ^ s ^ "?" ] in
-    k s
+    sk (k s)
 ;;
 
 let runConsole output =
   let open IO in
   let i c = runConsoleIO output c in
-  { interpreter = i }
+  { handler = i }
 ;;
 
 let write_hello () =

--- a/test/preface_examples_test/freer_monad_os_effect.ml
+++ b/test/preface_examples_test/freer_monad_os_effect.ml
@@ -1,0 +1,130 @@
+let try_testable a =
+  Alcotest.testable
+    (Preface.Try.pp (Alcotest.pp a))
+    (Preface.Try.equal (Alcotest.equal a))
+;;
+
+type 'a effect =
+  | Get_home : string effect
+  | Print : string -> unit effect
+
+module Effect = struct
+  include Preface.Make.Freer_monad.Over (struct
+    type 'a t = 'a effect
+  end)
+
+  let get_home = perform Get_home
+
+  let print message = perform (Print message)
+
+  let program path =
+    let open Monad in
+    match path with
+    | Some path ->
+      let* () = print ("Path is " ^ path) in
+      return (Preface.Try.ok ())
+    | None ->
+      let* home = get_home in
+      let* () = print ("Path is " ^ home) in
+      return (Preface.Try.ok ())
+  ;;
+end
+
+exception No_home
+
+let record reference action = reference := !reference @ [ action ]
+
+let happy_path_without_path () =
+  let output = ref [] in
+  let expected = [ "get_home"; "print Path is /xhtmlboi" ] in
+  let result =
+    let open Effect in
+    run
+      {
+        handler =
+          (fun resume effect ->
+            let f : type b. (b -> 'a) -> b effect -> 'a =
+             fun resume -> function
+              | Print message ->
+                let () = record output ("print " ^ message) in
+                resume ()
+              | Get_home ->
+                let () = record output "get_home" in
+                resume "/xhtmlboi"
+            in
+            f resume effect)
+      }
+      (program None)
+  in
+  Alcotest.(check (try_testable unit)) "should be equal" result (Ok ());
+  Alcotest.(check (list string)) "should be equal" expected !output
+;;
+
+let happy_path_with_path () =
+  let output = ref [] in
+  let expected = [ "print Path is ./a-path" ] in
+  let result =
+    let open Effect in
+    run
+      {
+        handler =
+          (fun resume effect ->
+            let f : type b. (b -> 'a) -> b effect -> 'a =
+             fun resume -> function
+              | Print message ->
+                let () = record output ("print " ^ message) in
+                resume ()
+              | Get_home ->
+                let () = record output "get_home" in
+                resume "/xhtmlboi"
+            in
+            f resume effect)
+      }
+      (program (Some "./a-path"))
+  in
+  Alcotest.(check (try_testable unit)) "should be equal" result (Ok ());
+  Alcotest.(check (list string)) "should be equal" expected !output
+;;
+
+let unhappy_path_without_path () =
+  let output = ref [] in
+  let expected = [ "get_home" ] in
+  let result =
+    let open Effect in
+    run
+      {
+        handler =
+          (fun resume effect ->
+            let f : type b. (b -> 'a) -> b effect -> 'a =
+             fun resume -> function
+              | Print message ->
+                let () = record output ("print " ^ message) in
+                resume ()
+              | Get_home ->
+                let () = record output "get_home" in
+                Preface.Try.error No_home
+            in
+            f resume effect)
+      }
+      (program None)
+  in
+  Alcotest.(check (try_testable unit))
+    "should be equal" result
+    (Preface.Try.error No_home);
+  Alcotest.(check (list string)) "should be equal" expected !output
+;;
+
+let cases =
+  [
+    ( "Freer Monad OS effect"
+    , let open Alcotest in
+      [
+        test_case "Happy path: perform program without path" `Quick
+          happy_path_without_path
+      ; test_case "Happy path: perform program with path" `Quick
+          happy_path_with_path
+      ; test_case "Unhappy path: perform program without path" `Quick
+          unhappy_path_without_path
+      ] )
+  ]
+;;

--- a/test/preface_examples_test/freer_monad_os_explicit_continuation.ml
+++ b/test/preface_examples_test/freer_monad_os_explicit_continuation.ml
@@ -1,0 +1,118 @@
+let try_testable a =
+  Alcotest.testable
+    (Preface.Try.pp (Alcotest.pp a))
+    (Preface.Try.equal (Alcotest.equal a))
+;;
+
+type 'a effect =
+  | Get_home of (string -> 'a)
+  | Print of (string * (unit -> 'a))
+
+module Effect = struct
+  include Preface.Make.Freer_monad.Over (struct
+    type 'a t = 'a effect
+  end)
+
+  let get_home = perform (Get_home Fun.id)
+
+  let print message = perform (Print (message, Fun.id))
+
+  let program path =
+    let open Monad in
+    match path with
+    | Some path ->
+      let* () = print ("Path is " ^ path) in
+      return (Preface.Try.ok ())
+    | None ->
+      let* home = get_home in
+      let* () = print ("Path is " ^ home) in
+      return (Preface.Try.ok ())
+  ;;
+end
+
+exception No_home
+
+let record reference action = reference := !reference @ [ action ]
+
+let happy_path_without_path () =
+  let output = ref [] in
+  let expected = [ "get_home"; "print Path is /xhtmlboi" ] in
+  let result =
+    let open Effect in
+    run
+      {
+        handler =
+          (fun resume -> function
+            | Print (message, k) ->
+              let () = record output ("print " ^ message) in
+              resume (k ())
+            | Get_home k ->
+              let () = record output "get_home" in
+              resume (k "/xhtmlboi"))
+      }
+      (program None)
+  in
+  Alcotest.(check (try_testable unit)) "should be equal" result (Ok ());
+  Alcotest.(check (list string)) "should be equal" expected !output
+;;
+
+let happy_path_with_path () =
+  let output = ref [] in
+  let expected = [ "print Path is ./a-path" ] in
+  let result =
+    let open Effect in
+    run
+      {
+        handler =
+          (fun resume -> function
+            | Print (message, k) ->
+              let () = record output ("print " ^ message) in
+              resume (k ())
+            | Get_home k ->
+              let () = record output "get_home" in
+              resume (k "/xhtmlboi"))
+      }
+      (program (Some "./a-path"))
+  in
+  Alcotest.(check (try_testable unit)) "should be equal" result (Ok ());
+  Alcotest.(check (list string)) "should be equal" expected !output
+;;
+
+let unhappy_path_without_path () =
+  let output = ref [] in
+  let expected = [ "get_home" ] in
+  let result =
+    let open Effect in
+    run
+      {
+        handler =
+          (fun resume -> function
+            | Print (message, k) ->
+              let () = record output ("print " ^ message) in
+              resume (k ())
+            | Get_home _ ->
+              let () = record output "get_home" in
+              Preface.Try.error No_home)
+      }
+      (program None)
+  in
+  Alcotest.(check (try_testable unit))
+    "should be equal" result
+    (Preface.Try.error No_home);
+  Alcotest.(check (list string)) "should be equal" expected !output
+;;
+
+let cases =
+  [
+    ( "Freer Monad OS effect with explicit continuation"
+    , let open Alcotest in
+      [
+        test_case "Happy path: perform program without path" `Quick
+          happy_path_without_path
+      ; test_case "Happy path: perform program with path" `Quick
+          happy_path_with_path
+      ; test_case "Unhappy path: perform program without path" `Quick
+          unhappy_path_without_path
+      ] )
+  ]
+;;

--- a/test/preface_examples_test/preface_example_test.ml
+++ b/test/preface_examples_test/preface_example_test.ml
@@ -6,6 +6,8 @@ let () =
     @ Shape.cases
     @ Free_monad_consoleIO.cases
     @ Freer_monad_consoleIO.cases
+    @ Freer_monad_os_effect.cases
+    @ Freer_monad_os_explicit_continuation.cases
     @ Template_reader.cases
     @ Debruijn_reader.cases )
 ;;

